### ISSUE TITLE
Unused interface property in an object spread

### ISSUE THREAD

**Issue by bodinsamuel** [microsoft/TypeScript#36259]
## Search Terms

unused required type definition

## Suggestion
Throw a warning (or suggest) if a required property in interface is never used in an object spread.

Right now, it does not warn because I suppose the spread of an object is still an acceptable subset of the declared interface, but it will throw on usage so the dev will likely try to respect the interface whereas the code will not make use of it. 
This can lead to making the wrong assumption or confusion about the code.

This can be part of a "strict" rule.

N.B: I'm not sure if it's a breaking change or not, **this also could be implemented in eslint,**
please feel free to dismiss this proposal if so.

## Use Cases
Proactively detect dead code.

## Examples
```ts
// no warning
function myFunc({ foo }: { foo: string; bar: number }) {
  return foo;
}

// myFunc will throw a warning because "bar" is required
myFunc({ foo: 'hello' });
````

Playground link: https://www.typescriptlang.org/play/?useDefineForClassFields=true&noUnusedLocals=true&noUnusedParameters=true&preserveConstEnums=true&removeComments=true&checkJs=true&allowJs=true&experimentalDecorators=true&emitDecoratorMetadata=true#code/GYVwdgxgLglg9mABAWwJ4DFwQBQG9HBxyIC+AXIvoXBQM5QBOMYA5gNyIBGAhgxWCGScApg1IBKSgFgAUIkQNhUEAyTU2skrNlpMkPASIUA5AAthAGwtxjEtkA

**Code that could lead to confusion**
```ts
function myFunc({ foo }: { foo: string; callback: () => void }) {
  return foo;
}

// the developper will fill callback thinking the outcome will come in callback
myFunc({ 
  foo: 'hello', 
  callback: () => {
  } 
});
````

## Checklist

My suggestion meets these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [*] This wouldn't change the runtime behavior of existing JavaScript code
* [*] This could be implemented without emitting different JS based on the types of the expressions
* [*] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [*] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by RyanCavanaugh** [comment]
This isn't indicative of a type error. For example, you might want to take an array-like but reject a `window`:
```ts
function something({ length }: { length: number, push: any }) {
  // something
}
```

It would also be very weirdly inconsistent to say this this code is wrong
```ts
function myFunc({ foo }: { foo: string; bar: number }) {
  return foo;
}
```
but that this code is OK
```ts
function myFunc(f: { foo: string; bar: number }) {
  return f.foo;
}
```

Overall this does feel more like a lint rule since it makes fairly strong guesses about what the intent of code might be.

**Comment by bodinsamuel** [comment]
I guess you are right it doesn't work if you require a complete interface to be passed for the validation, but only use part of it.
This is only "good" for object spread and inline interface, and even there it might be arguable depending on the use case.

Fair enough 👍
